### ISSUE TITLE
Custom YAML tag processors

### ISSUE THREAD

**Issue by Blacksmoke16** [crystal-lang/crystal#8598]
Currently the YAML parser handles a static set of [tags](https://github.com/crystal-lang/crystal/blob/master/src/yaml/schema/core.cr#L275-L292).  I propose to allow custom handlers to be registered to allow for handling custom tags.

This would also allow the stdlib to register some utility tags, like `env: !crystal/env APP_ENV` for example.

**Comment by straight-shoota** [comment]
This should be pretty easy, just needs to hook into `process_tag`/`process_scalar_tag`.

But I'm not sure whether tags should be registered globally or specific to a `YAML::Parser` instance. Maybe even both?

**Comment by ryanprior** [comment]
Globally would not be necessary, you might want to parse two different kinds of tagged YAML in a single program & should not have to scratch your head to do that :)

**Comment by Sija** [comment]
@ryanprior Yeah, but that doesn't mean you wouldn't like to have some (inheritable?) defaults.

**Comment by asterite** [comment]
This is easy to implement is the handler creates one of the types supported by YAML.parse, like int, string, time, etc., but I have no idea how it would work with custom types...

**Comment by Blacksmoke16** [comment]
> But I'm not sure whether tags should be registered globally or specific to a `YAML::Parser` instance. Maybe even both?

Is there a reason I'm not seeing that would make it better to not have it global?

> Globally would not be necessary, you might want to parse two different kinds of tagged YAML in a single program & should not have to scratch your head to do that :)

I would assume that if a YAML document doesn't have any of a registered tag, it just doesn't do anything like now.  Unless you mean processing the same tag differently in different files?

> This is easy to implement is the handler creates one of the types supported by YAML.parse, like int, string, time, etc., but I have no idea how it would work with custom types...


It should also ideally work with `YAML::Serializable`.  I want this feature for some Athena stuff so I'll work on coming up with some implementation options.

**Comment by konovod** [comment]
> Is there a reason I'm not seeing that would make it better to not have it global?

If application loads YAML config from trusted source (and can use `env` tag there) and then use YAML to parse some untrusted info (and `env` shouldn't be allowed there).

**Comment by Blacksmoke16** [comment]
Based on what I learned from making the prototype and the information from https://github.com/crystal-lang/crystal/pull/8616#issuecomment-568629481, here's my proposal:
* Do not provide anything custom processors by default
* Have a way to define processors that would handle _all_ documents
* Have an option that would allow you to pass the processors you wish to use to `YAML.parse` or `Type.from_yaml`
  * These processors could override/be merged with (TBD) the global ones.
  * A processor could be a concrete type, like a struct that implements a `#process` method or something.
    * `MyTagHandler.new`
  * Processors could also just be a proc like it is now
    * Or support both?

Ex.

```cr
record MyTagHandler < YAML::TagHandler do
  def tag : String
    "!myTag"
  end

  def process(value : String)
    # Do your logic here
  end
end
```



\cc @lbguilherme

**Comment by Blacksmoke16** [comment]
Actually maybe a `!crystal/regex` would be a good tag. Would allow parsing a string as `Regex`, which currently isn't possible. Has to be a tag to prevent non regex strings from being converted to regex as well.

**Comment by Sija** [comment]
@Blacksmoke16 Having such in stdlib - sure! Havin' it on by default - plz, no, see https://github.com/crystal-lang/crystal/pull/8616#issuecomment-568629481

**Comment by straight-shoota** [comment]
Let's postpone the discussion about whether and which tags to ship with stdlib and/or enable by default. That's only the second step. We should focus on the general implementation first.

* A benefit of using a custom handler type per tag is that this could probably be used with `YAML::Any`. We could add a module to `YAML::Any::Type` and tag types would need to implement that module. Procs won't work for that.
* In order to keep this as small as possible, I'd consider to maybe not even use procs at all.

You could also take a look at `Psych.add_tag` and .add_domain_type` from Ruby's stdlib which would be pretty similar to what you're suggesting. Looking at implementations in other languages could be useful, too.

**Comment by RX14** [comment]
tags and their meanings should surely be bound to a `YAML::ParseContext`.

Figure out the core abstracton before worrying about `Any` or any of the things built on top of the raw parser.
